### ISSUE TITLE
Broken library ?

### ISSUE THREAD

**Issue by Weakky** [Eastrall/Ether.Network#45]
Hello sir,

There are two issues here.

With version `2.1.0-pre`, the Client is completely stuck after sending one message. (using the readme.md example). No errors are thrown, it's just waiting.

As for version `2.0.1`, the client doesn't freeze but when flooding the server using the sample client, it ends up quickly by crashing, raising this exception:

```
Unhandled Exception: System.IO.EndOfStreamException: Unable to read beyond the end of the stream.
   at System.IO.__Error.EndOfFile()
   at System.IO.MemoryStream.InternalReadInt32()
   at Ether.Network.Packets.NetPacketMethods.<>c.<.cctor>b__2_13(BinaryReader reader)
   at Ether.Network.Packets.NetPacketBase.Read[T]()
   at CardGame.Server.ClientConnection.HandleMessage(NetPacketBase packet) in /Users/flavian/Projects/DotNet/CardGame/CardGame.Server/ClientConnection.cs:line 30
   at Ether.Network.NetServer`1.DispatchPackets(NetConnection netConnection, SocketAsyncEventArgs e)
   at Ether.Network.NetServer`1.ProcessReceive(SocketAsyncEventArgs e)
   at System.Net.Sockets.SocketAsyncEventArgs.OnCompleted(SocketAsyncEventArgs e)
   at System.Net.Sockets.SocketAsyncEventArgs.ExecutionCallback(Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncSuccess(Int32 bytesTransferred, SocketFlags flags)
   at System.Net.Sockets.SocketAsyncEventArgs.CompletionCallback(Int32 bytesTransferred, SocketFlags flags, SocketError socketError)
   at System.Net.Sockets.SocketAsyncEventArgs.TransferCompletionCallbackCore(Int32 bytesTransferred, Byte[] socketAddress, Int32 socketAddressSize, SocketFlags receivedFlags, SocketError socketError)
   at System.Net.Sockets.SocketAsyncContext.ReceiveOperation.InvokeCallback()
   at System.Net.Sockets.SocketAsyncContext.AsyncOperation.<>c.<TryCompleteOrAbortAsync>b__14_0(Object o)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**The sample example works correctly though.**

Do you have any ideas how I could fix this ?

Thanks for help.

**Comment by Weakky** [comment]
By the way, ain't there any way to broadcast a message ?

**Comment by Er3van** [comment]
Mmmh, I have no issue in sending a large amount of data from client to server, can you post or describe what you did exactly to reproduce this error, would be good.

**Comment by Er3van** [comment]
I'm looking for a way to broadcast a message too BTW @Eastrall

**Comment by Weakky** [comment]
I have currently solved the issue by cloning the whole repo, imported it in the project and set it as a dependency. I'm currently using the version on the `develop` branch, and it works well. The code used was the one found in the samples.

I'm currently looking deeper in the implementation and it seems there are no way to do this atm. I may be wrong though.

**Comment by Er3van** [comment]
I took a look too, and except adding a member variable for each connection and setting a message in it, with a bool saying it has to be broadcasted, then looping over the clients in a thread in the server and broadcasting the message to all the users if the variable is set, I have no other idea in mind

BTW, we are all from Epitech and I think @Weakky you're on CardGame too ^^

**Comment by Weakky** [comment]
Haha. In order not to flood the issue here, could you please reach me at the email you can find on my profile so we can discuss about that ?

**Comment by Eastrall** [comment]
Hello,

Thank you for your feedback about this bug. In fact, this issue has already been reported : https://github.com/Eastrall/Ether.Network/issues/35
The version `2.1.0-pre` appeared to be an unstable version since I changed the behavior of the package manager. I suggest you to use the version `2.0.1`.

If you want to broadcast a message to all your clients, you can use the `NetConnection. SendTo(ICollection<NetConnection> clients, NetPacketBase packet)` method.

If you use the `develop` branch, be aware that this branch is also unstable since it is in constant development and might have unreported bugs.

**Comment by Weakky** [comment]
I see,wonderful. The reason why I didn't have access to `NetConnection.SendTo` method is because you moved it into the `NetConnectionOld` file. I checked out your fix and it's only a few lines so I'm just gonna use version `2.0.1` and apply the fix to it. Thank you very much for your help sir.

**Comment by Eastrall** [comment]
Didn't removed the `NetConnectionOld` file yet, since I have done a smal refactoring to the `NetConnection` and added the `NetUser` class. I have to remove it.

You're welcome, good luck with your project.
Be aware tho, that I am not responsible if you get a bad mark since this project is an open-source/community project always evolving.

**Comment by Eastrall** [comment]
In version `2.0.1`, yes you need to pass a reference of the `NetServer<T>` to your clients. Not great I agree. 

In version `3.0` I will try to implement a method to the `INetServer` interface that can broadcast a message to all connected clients or only to a list, so instead of passing the `NetServer<T>` implementation to your clients, you can only pass the `INetServer` interface.
Or maybe make a `protected` field in the `NetUser` class. Don't really know yet.

**Comment by Eastrall** [comment]
Most of the errors have been fixed for version `3.0.0` in PR #49 and #50.
Receive and sending processes have been improved and tested.
Next feature to implement, the "broadcast" system to send a message to all clients connected.

**Comment by Eastrall** [comment]
Broadcast system has been added in PR #52.

In version 3.0.0, you can now broadcast a message using the `INetServer.SendToAll(INetPacketStream)` or `INetServer.SendTo(IEnumerable<INetUser>, INetPacketStream)`.
